attestation-container/attest/attestation_endorsement.go

```diff
@@ -0,0 +1,118 @@
+package attest
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"math"
+	"math/rand"
+	"net/http"
+	"time"
+)
+
+const (
+	defaultBaseSec    = 2
+	defaultMaxRetries = 5
+)
+
+func fetchWithRetry(requestURL string, baseSec int, maxRetries int) ([]byte, error) {
+	if maxRetries < 0 {
+		return nil, fmt.Errorf("invalid `maxRetries` value")
+	}
+	var err error
+	retryCount := 0
+	for retryCount <= maxRetries {
+		if retryCount > 0 {
+			// Exponential backoff
+			maxDelay := math.Pow(float64(baseSec), float64(retryCount))
+			delaySec := rand.Float64() * maxDelay
+			delaySecInt := math.Min(math.MaxInt64, delaySec)
+			time.Sleep(time.Duration(delaySecInt) * time.Second)
+		}
+		res, err := http.Get(requestURL)
+		if err != nil {
+			retryCount++
+			continue
+		}
+		if 200 <= res.StatusCode && res.StatusCode < 300 {
+			// Got successful status code 2xx
+			defer res.Body.Close()
+			resBody, err := ioutil.ReadAll(res.Body)
+			if err != nil {
+				retryCount++
+				continue
+			}
+			return resBody, nil
+		} else if res.StatusCode == 408 || res.StatusCode == 429 || 500 <= res.StatusCode {
+			// Got status code that is worth to retry
+			retryCount++
+			continue
+		} else {
+			// Got status code that is not worth to retry
+			defer res.Body.Close()
+			resBody, err := ioutil.ReadAll(res.Body)
+			if err != nil {
+				return nil, fmt.Errorf("got error while handling non successful response with status code %d: %s", res.StatusCode, err)
+			}
+			return nil, fmt.Errorf("GET request failed with status code %d: %s", res.StatusCode, resBody)
+		}
+	}
+	return nil, err
+}
+
+func fetchAttestationEndorsementAzure(reportedTCBBytes [REPORTED_TCB_SIZE]byte, chipID string) ([]byte, error) {
+	// Fetch attestation endorsement from Azure endpoint
+	reportedTCB := binary.LittleEndian.Uint64(reportedTCBBytes[:])
+	reportedTCBHex := fmt.Sprintf("%x", reportedTCB)
+	requestURL := fmt.Sprintf("https://global.acccache.azure.net/SevSnpVM/certificates/%s/%s?api-version=2020-10-15-preview", chipID, reportedTCBHex)
```

Comment:
I feel like the Azure and AMD endpoint addresses should be moved to the top of the file as global constants?

Reply:
[51b028d](https://github.com/microsoft/CCF/pull/4795/commits/51b028d94292869e9f69a8d82b354b781be01785)